Content.Server/GameObjects/EntitySystems/Atmos/AtmosphereSystem.Devices.cs

```diff
@@ -0,0 +1,148 @@
+        }
+
+        public void JoinAtmosphere(AtmosDeviceComponent component)
+        {
+            if (!CanJoinAtmosphere(component))
+                return;
+
+            // We try to get a valid, simulated atmosphere.
+            if (!TryGetSimulatedGridAtmosphere(component.Owner.Transform.MapPosition, out var atmosphere))
+                return;
+
+            component.Atmosphere = atmosphere;
+            atmosphere.AddAtmosDevice(component);
+        }
+
+        public void LeaveAtmosphere(AtmosDeviceComponent component)
+        {
+            component.Atmosphere?.RemoveAtmosDevice(component);
+            component.Atmosphere = null;
+        }
+
+        public void RejoinAtmosphere(AtmosDeviceComponent component)
+        {
+            LeaveAtmosphere(component);
+            JoinAtmosphere(component);
+        }
+
+        private void OnDeviceInitialize(EntityUid uid, AtmosDeviceComponent component, ComponentInit args)
+        {
+            JoinAtmosphere(component);
+        }
+
+        private void OnDeviceShutdown(EntityUid uid, AtmosDeviceComponent component, ComponentShutdown args)
+        {
+            LeaveAtmosphere(component);
+        }
+
+        private void OnDeviceBodyTypeChanged(EntityUid uid, AtmosDeviceComponent component, PhysicsBodyTypeChangedEvent args)
+        {
+            // Do nothing if the component doesn't require being anchored to function.
+            if (!component.RequireAnchored)
+                return;
+
+            if (args.New == BodyType.Static)
+                JoinAtmosphere(component);
+            else
+                LeaveAtmosphere(component);
+        }
+
+        private void OnDeviceParentChanged(EntityUid uid, AtmosDeviceComponent component, EntParentChangedMessage args)
+        {
+            RejoinAtmosphere(component);
+        }
+        #endregion
+
+        #region UnsafeUnanchor
+        private void OnUnanchorAttempt(EntityUid uid, AtmosUnsafeUnanchorComponent component, UnanchorAttemptEvent args)
+        {
+            if (!component.Enabled || !ComponentManager.TryGetComponent(uid, out NodeContainerComponent? nodes))
+                return;
+
+            if (!component.Owner.Transform.Coordinates.TryGetTileAir(out var environment, EntityManager))
+                return;
+
+            foreach (var node in nodes.Nodes.Values)
+            {
+                if (node is not PipeNode pipe) continue;
+
+                if ((pipe.Air.Pressure - environment.Pressure) > 2 * Atmospherics.OneAtmosphere)
+                {
+                    args.Delay += 1.5f;
+                    args.User?.PopupMessageCursor(Loc.GetString("comp-atmos-unsafe-unanchor-warning"));
+                    return; // Show the warning only once.
+                }
+            }
+        }
+
+        private void OnBeforeUnanchored(EntityUid uid, AtmosUnsafeUnanchorComponent component, BeforeUnanchoredEvent args)
```

Comment:
Awww... you didn't add my "yeet the person backwards for being stupid" :(

bring back `fuck_you_dir` ;_____;

Reply:
I'll add a TODO, then :relieved: